src/main/scala/com/fulcrumgenomics/commons/util/Async.scala

```diff
@@ -0,0 +1,286 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.commons.util
+
+import java.io.Closeable
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger, AtomicReference}
+import java.util.concurrent.{ArrayBlockingQueue, BlockingQueue, TimeUnit}
+
+import com.fulcrumgenomics.commons.io.Writer
+
+import scala.concurrent.forkjoin.ForkJoinPool
+
+/**
+  * Provides generic utility classes related to asynchronous processing, reading, writing, and iterating over items.
+  */
+object Async {
+
+  /** The default # of items to buffer in queues. */
+  private val DefaultBufferSize: Int = 2048
+
+  private object AsyncIterator {
+    /** A [[Runnable]] who can be asked when it is all done.
+      *
+      * @param f the method to run
+      */
+    case class Runner(f: () => Unit) extends Runnable {
```

Comment:
I don't think this class is need and it obfuscates what's going on.  In can just have the `allDone` member inside `AsyncIterator` and declare runnable as:

```scala
val runnable = () => { outer.source.foreach(outer.queue.put); allDone.set(true) }
```